modules/twinkleblock.js

```diff
@@ -0,0 +1,870 @@
+    $(el).prop('checked', check);
+  });
+  $form.find('[name=reason]').val(data.reason);
+};
+
+Twinkle.block.callback.change_subcategory = function twinkleblockcallbackChangeSubcategory(e) {
+  var form = e.target.form, prev_block_timer = Twinkle.warn.prev_block_timer;
+  form.main_group.value = 'block';
+  Twinkle.warn.callback.change_subcategory(e);
+  if (prev_block_timer) form.expiry.value = prev_block_timer;
+};
+
+Twinkle.block.callback.evaluate = function twinkleblockCallbackEvaluate(e) {
+  var $form = $(e.target),
+      toBlock = $form.find('[name=actiontype][value=block]').is(':checked'),
+      toWarn = $form.find('[name=actiontype][value=template]').is(':checked');
+
+  if (!toBlock && !toWarn) return alert("Please give Twinkle something to do!");
+
+  if (toBlock) {
+    var blockoptions = $form.find('[name=field_block_options]').serializeObject(),
+        warnoptions = $form.find('[name=field_template_options]').serializeObject();
+    delete blockoptions.expiry_preset; // remove extraneous
+
+    if (!blockoptions.expiry) return alert('Please provide an expiry!');
+    if (blockoptions.reason || (!blockoptions.reason && confirm('No block reason was provided. Continue?'))) {
+      Morebits.simpleWindow.setButtonsEnabled( false );
+      Morebits.status.init( e.target );
+      formCleared = true;
+      blockoptions.action = 'block';
+      blockoptions.user = relevantUser;
+
+      // boolean-flipped options
+      blockoptions.anononly = blockoptions.hardblock ? undefined : true;
+      blockoptions.allowusertalk = blockoptions.disabletalk ? undefined : true;
+
+      // temporary fix for bug with block API, see [[phab:T68646]]
+      if (blockoptions.expiry === 'infinity') blockoptions.expiry = 'infinite';
+
+      // execute block
+      api.postWithToken('block', blockoptions).then(function(data) {
+        Morebits.status.info('Info', 'Blocked ' + relevantUserName);
+        if (toWarn) Twinkle.block.callback.warn_user(e, warnoptions);
+      }, function(msg, data) {
+        Morebits.status.warn('Error', 'Failed to block ' + relevantUserName);
+        Morebits.status.warn('Reason', data.error.info);
+      });
+    }
+  } else if (toWarn) {
+    Twinkle.block.callback.warn_user(e);
+  }
+};
+
+Twinkle.block.callback.warn_user = function twinkleblockCallbackWarnUser(e, serializedForm) {
+  var form = e.target.block_reason ? e.target : serializedForm,
+      userTalkPage = "User_talk:" + relevantUser;
+
+  // set up params needed by warn module, serialized form data will not have value property
+  var params = {
+    reason: form.block_reason.value || form.block_reason,
+    main_group: 'block',
+    sub_group: form.sub_group.value || form.sub_group,
+    article: form.article.value || form.article,
+    block_timer: form.block_timer.value || form.block_timer,
+    messageData: Twinkle.block.blockPresetsInfo[form.sub_group.value || form.sub_group]
+  };
+
+  if (!formCleared) {
```

Comment:
I haven't closely followed the chain of logic related to `formCleared` yet, but no other Twinkle modules require this, so I am a bit sceptical of it.